### PR TITLE
Add max_single_primary_size as a condition for the ILM rollover action

### PR DESCRIPTION
**Pull request by joegallo**
Related to #63026, followup to #67842

Threads the new `max_single_primary_size` rollover trigger from #67842 into ILM's `RolloverAction`.

### CODE REVIEW

**Comment by joegallo** [comment]
IMHO reviewing commit by commit would be best, but that's just me.

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ilm/RolloverAction.java [hunk 1]
```diff
@@ -34,35 +35,45 @@
...
 
-    public RolloverAction(@Nullable ByteSizeValue maxSize, @Nullable TimeValue maxAge, @Nullable Long maxDocs) {
-        if (maxSize == null && maxAge == null && maxDocs == null) {
+    public RolloverAction(@Nullable ByteSizeValue maxSize, @Nullable ByteSizeValue maxSinglePrimarySize, @Nullable TimeValue maxAge,
+                          @Nullable Long maxDocs) {
+        if (maxSize == null && maxSinglePrimarySize == null && maxAge == null && maxDocs == null) {
             throw new IllegalArgumentException("At least one rollover condition must be set.");
         }
         this.maxSize = maxSize;
+        this.maxSinglePrimarySize = maxSinglePrimarySize;
```
Comment (andreidan) [inline x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ilm/RolloverAction.java:76]: would it make sense to validate that `maxSinglePrimarySize` is let to `maxSize`?

File: x-pack/plugin/ilm/qa/multi-node/src/javaRestTest/java/org/elasticsearch/xpack/ilm/TimeSeriesLifecycleActionsIT.java [hunk 2]
```diff
@@ -223,6 +228,40 @@ public void testRolloverActionWithIndexingComplete() throws Exception {
...
+        List<Integer> primaryShardSizes = getPrimaryShardSizes(client(), originalIndex);
+        Integer largest = Collections.max(primaryShardSizes);
+
+        // create policy
+        createNewSingletonPolicy(client(), policy, "hot", new RolloverAction(
+            null,
+            // the largest size we have from above is index stats size, but that's larger
+            // than the docstats size that we actually evaluate against inside the guts
+            // of rollover, 2048 is a slop factor
+            ByteSizeValue.ofBytes(Long.max(1, largest - 2048)),
```
Comment (andreidan) [inline x-pack/plugin/ilm/qa/multi-node/src/javaRestTest/java/org/elasticsearch/xpack/ilm/TimeSeriesLifecycleActionsIT.java:254]: I am inclined to say this sort of complexity is not needed here. We're not meant to test the accuracy of shard size stats here, but that the condition is triggered. 
Shall we just use `1` here and also delete `getPrimaryShardSizes`? 

What do you think?

File: x-pack/plugin/ilm/qa/multi-node/src/javaRestTest/java/org/elasticsearch/xpack/ilm/TimeSeriesLifecycleActionsIT.java [hunk 3]
```diff
@@ -223,6 +228,40 @@ public void testRolloverActionWithIndexingComplete() throws Exception {
             getOnlyIndexSettings(client(), originalIndex).get(LifecycleSettings.LIFECYCLE_INDEXING_COMPLETE)));
     }
 
+    public void testRolloverActionWithMaxSinglePrimarySize() throws Exception {
```
Comment (andreidan) [inline x-pack/plugin/ilm/qa/multi-node/src/javaRestTest/java/org/elasticsearch/xpack/ilm/TimeSeriesLifecycleActionsIT.java:231]: `TimeSeriesLifecycleActionsIT` has become quite big and we're moving away from adding to it, towards having one IT/action ( https://github.com/elastic/elasticsearch/issues/53488 )
We can leave this test here for now, but would be awesome to separate the `rollover` tests into a `RolloverActionIT` (in a future PR maybe)

File: x-pack/plugin/ilm/qa/multi-node/src/javaRestTest/java/org/elasticsearch/xpack/ilm/TimeSeriesLifecycleActionsIT.java [hunk 4]
```diff
@@ -223,6 +228,40 @@ public void testRolloverActionWithIndexingComplete() throws Exception {
...
+            // the largest size we have from above is index stats size, but that's larger
+            // than the docstats size that we actually evaluate against inside the guts
+            // of rollover, 2048 is a slop factor
+            ByteSizeValue.ofBytes(Long.max(1, largest - 2048)),
+            null, null));
+        // update policy on index
+        updatePolicy(client(), originalIndex, policy);
+
+        assertBusy(() -> assertTrue(indexExists(secondIndex)));
+        assertBusy(() -> assertTrue(indexExists(originalIndex)));
```
Comment (andreidan) [inline x-pack/plugin/ilm/qa/multi-node/src/javaRestTest/java/org/elasticsearch/xpack/ilm/TimeSeriesLifecycleActionsIT.java:260]: due to various flakiness issues we usually put a 30 seconds timeout on these as CI might be slow (it defaults to 10 seconds I think)

File: x-pack/plugin/ilm/qa/multi-node/src/javaRestTest/java/org/elasticsearch/xpack/ilm/TimeSeriesLifecycleActionsIT.java [hunk 5]
```diff
@@ -223,6 +228,40 @@ public void testRolloverActionWithIndexingComplete() throws Exception {
...
+        assertBusy(() -> assertEquals("true",
+            getOnlyIndexSettings(client(), originalIndex).get(LifecycleSettings.LIFECYCLE_INDEXING_COMPLETE)));
+    }
```
Comment (andreidan) [inline x-pack/plugin/ilm/qa/multi-node/src/javaRestTest/java/org/elasticsearch/xpack/ilm/TimeSeriesLifecycleActionsIT.java:263]: shall we also add an assertion that ILM completed for the original index? (basically a "leave things in order" assertion :) )

ie. ```
        assertBusy(() -> assertThat(getStepKeyForIndex(client(), originalIndex), equalTo(PhaseCompleteStep.finalStep("hot").getKey())));
```

**Review by andreidan — COMMENTED** [review]
Thanks for working on this Joe. This generally looks great! Nicely done.

I left a few rather minor suggestions and questions.

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ilm/RolloverAction.java (hunk 1 above)
Comment (joegallo) [reply x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ilm/RolloverAction.java:76]: IMHO no -- we don't sanity check the other things (e.g. `maxDocs` 1 paired with `maxSize` 10,000GB is just as valid as `maxDocs` 10,000 with `maxSize` 1 byte -- at least AFAICT), so I think we should be similarly lenient with `maxSinglePrimarySize`.

File: x-pack/plugin/ilm/qa/multi-node/src/javaRestTest/java/org/elasticsearch/xpack/ilm/TimeSeriesLifecycleActionsIT.java (hunk 3 above)
Comment (joegallo) [reply x-pack/plugin/ilm/qa/multi-node/src/javaRestTest/java/org/elasticsearch/xpack/ilm/TimeSeriesLifecycleActionsIT.java:231]: +1, but will do as a separate PR

File: x-pack/plugin/ilm/qa/multi-node/src/javaRestTest/java/org/elasticsearch/xpack/ilm/TimeSeriesLifecycleActionsIT.java (hunk 2 above)
Comment (joegallo) [reply x-pack/plugin/ilm/qa/multi-node/src/javaRestTest/java/org/elasticsearch/xpack/ilm/TimeSeriesLifecycleActionsIT.java]: 4393939

File: x-pack/plugin/ilm/qa/multi-node/src/javaRestTest/java/org/elasticsearch/xpack/ilm/TimeSeriesLifecycleActionsIT.java (hunk 4 above)
Comment (joegallo) [reply x-pack/plugin/ilm/qa/multi-node/src/javaRestTest/java/org/elasticsearch/xpack/ilm/TimeSeriesLifecycleActionsIT.java]: 2ad066c9eb7cbd6dd922dcf68854fea61c5b933a

File: x-pack/plugin/ilm/qa/multi-node/src/javaRestTest/java/org/elasticsearch/xpack/ilm/TimeSeriesLifecycleActionsIT.java (hunk 5 above)
Comment (joegallo) [reply x-pack/plugin/ilm/qa/multi-node/src/javaRestTest/java/org/elasticsearch/xpack/ilm/TimeSeriesLifecycleActionsIT.java:253]: Also handled in 2ad066c9eb7cbd6dd922dcf68854fea61c5b933a

**Review by andreidan — APPROVED** [review]
LGTM, thanks Joe

**Comment by joegallo** [comment]
@elastic/es-ui there's a new option for ILM Rollovers (in 7.13 and 8.0.0). On this PR it was added as `max_single_primary_size` but we renamed it to `max_primary_shard_size` on #69239.

File: x-pack/plugin/ilm/qa/multi-node/src/javaRestTest/java/org/elasticsearch/xpack/ilm/TimeSeriesLifecycleActionsIT.java (hunk 3 above)
Comment (joegallo) [reply x-pack/plugin/ilm/qa/multi-node/src/javaRestTest/java/org/elasticsearch/xpack/ilm/TimeSeriesLifecycleActionsIT.java:232]: https://github.com/elastic/elasticsearch/pull/69295

**Pull request merged**